conjureup/juju.py

```diff
@@ -22,6 +22,38 @@
 PENDING_DEPLOYS = 0
 
 
+class JujuBinaryNotFound(Exception):
+    pass
+
+
+def set_bin_path():
+    """ Sets the juju binary path
+    """
+    if utils.is_darwin():
+        app.juju.bin_path = '/usr/local/bin/juju'
+    elif Path('/snap/bin/juju').exists():
+        app.juju.bin_path = '/snap/bin/juju'
+    elif Path('/snap/bin/conjure-up.juju').exists():
+        app.juju.bin_path = '/snap/bin/conjure-up.juju'
+    else:
+        raise JujuBinaryNotFound("Unable to locate a juju binary.")
```

Comment:
I think there are a few issues with this check.  We're not checking if the path actually exists on Mac OS, we're not checking if we can actually execute the files, and we don't support a non-snap install except on Mac OS (though, do we really want to?).  I would suggest the following instead:

```python
def set_bin_path():
    """ Sets the juju binary path
    """
    # search these paths, in order, for the juju binary
    # we don't use $PATH because we have definite preferences which one we use
    # and we don't want to leave it up to the user
    candidates = [
        '/snap/bin/juju',
        '/snap/bin/conjure-up.juju',
        '/usr/bin/juju',
        '/usr/local/bin/juju',
    ]
    for candidate in candidates:
        if os.access(candidate, os.X_OK):
            app.juju.bin_path = candidate
            app.log.debug("juju binary path: {}".format(app.juju.bin_path))
            break
    else:
        raise JujuBinaryNotFound("Unable to locate a juju binary.")
```